otp4gb/cost.py

```diff
@@ -0,0 +1,358 @@
+
+    gen_cost_column = f"{aggregation}_generalised_cost"
+
+    if gen_cost_column not in matrix.columns:
+        LOG.error(
+            "Generalised cost column (%s) not found in "
+            "matrix metrics, so skipping matrix creation",
+            gen_cost_column,
+        )
+        return
+
+    # Pivot generalised cost
+    gen_cost = matrix.pivot(
+        index="origin_id", columns="destination_id", values=gen_cost_column
+    )
+    gen_cost.to_csv(matrix_file)
+    LOG.info("Written %s generalised cost matrix to %s", aggregation, matrix_file)
+
+
+def build_cost_matrix(
+    jobs: list[CalculationParameters],
+    matrix_file: pathlib.Path,
+    generalised_cost_parameters: GeneralisedCostFactors,
+    aggregation_method: AggregationMethod,
+    workers: int = 0,
+) -> None:
+    """Create cost matrix for all zone to zone pairs.
+
+    Parameters
+    ----------
+    jobs : list[CalculationParameters]
+        List of requests to send to OTP.
+    matrix_file : pathlib.Path
+        Path to save the cost matrix to.
+    generalised_cost_parameters : GeneralisedCostParameters
+        Factors and other parameters for calculating the
+        generalised cost.
+    aggregation_method : AggregationMethod
+        Aggregation method used for generalised cost matrix.
+    workers : int, default 0
+        Number of threads to create during calculations.
+    """
+    LOG.info("Calculating costs for %s", matrix_file.name)
+
+    lock = threading.Lock()
+    response_file = matrix_file.with_name(matrix_file.name + "-response_data.jsonl")
+    with open(response_file, "wt", encoding=util.TEXT_ENCODING) as responses:
+        iterator = util.multithread_function(
+            workers,
+            calculate_costs,
+            jobs,
+            dict(
+                response_file=responses,
+                lock=lock,
+                generalised_cost_parameters=generalised_cost_parameters.copy(),
+            ),
+        )
+
+        matrix_data = []
+        for res in tqdm.tqdm(
+            iterator,
+            total=len(jobs),
+            desc="Calculating costs",
+            dynamic_ncols=True,
+            smoothing=0,
+        ):
+            matrix_data.append(res)
+
+    LOG.info("Written responses to %s", response_file)
+    _write_matrix_files(matrix_data, matrix_file, aggregation_method)
+
+
+def iterate_responses(response_file: io.TextIOWrapper) -> Iterator[CostResults]:
+    """Iterate through and parse responses JSON lines file.
+
+    Parameters
+    ----------
+    response_file : io.TextIOWrapper
+        JSON lines file to read from.
+
+    Yields
+    ------
+    CostResults
+        Cost results for a single OD pair.
+    """
+    for line in response_file:
+        yield CostResults.parse_raw(line)
+
+
+def cost_matrix_from_responses(
+    responses_files: list[pathlib.Path],
+    matrix_file: pathlib.Path,
+    aggregation_method: AggregationMethod,
+) -> None:
```

Comment:
Updated to positional arguments with [a27875a](https://github.com/Transport-for-the-North/otp4gb-py/commit/a27875a2fcde123ef7bb3310f7ee189ed2cef818)

Seems much more proper now. 

Thanks for all the examples above too which helped clarify it a bit.

One final Q: Will this allow me to pass a list of paths still if I require? 

Currently the [cost_metrics_from_responses.py](https://github.com/Transport-for-the-North/otp4gb-py/blob/OTP_v2_TRSE_/scripts/cost_metrics_from_responses.py) script takes a list of paths as a constant and then passes the entire list to the function.

i.e. would the following work
```
cost_matrix_from_responses(RESPONSES_PATHS,
                                              matrix_file=MATRIX_PATH,
                                              aggregation_method=AggregationMethod)
```

where RESPONSES_PATHS is a list of pathlib.Paths to each responses.jsonl

Reply:
you can't directly pass a list of parameters because that list would all be treated as a single argument so you would end up with a list with a single element containing your list but python has a nice way of splitting the list into separate arguments with `*`.

For your code to work you need to expand the list into separate arguments using `*` like below

```python
cost_matrix_from_responses(*RESPONSES_PATHS, matrix_file=MATRIX_PATH, aggregation_method=AggregationMethod)
```

Sidenote: you can use `**` to expand dictionaries into keyword arguments if needed so the code below would work (although I wouldn't use the code below but it's just as an example)

```python
cost_matrix_from_responses(
    *[path1, path2, path3], **{"matrix_file": MATRIX_PATH, "aggregation_method": AggregationMethod}
)
```
The above code effectively expands to the code below (although doing above method doesn't make sense because it's unecessary).

```python
cost_matrix_from_responses(
    path1, path2, path3, matrix_file=MATRIX_PATH, aggregation_method= AggregationMethod
)
```